### PR TITLE
various small fixes, improvements, updates

### PR DESCRIPTION
**Pull request by ngoldman**
Biggest change is a reorg of the index. Code is now grouped in terms of what it does. Should be a lot easier to read and parse (and refactor).

I also set up some logic for better plugin registration handling in the future. See the register, registerProvider, registerCache, and registerPlugin methods hanging off of `app`.

--

### Changed
* `lib/Files` now takes `options` instead of `koop` (only needed koop.config and koop.log)
* cleaned up `app.register` method logic
* using `koop.log` in index instead of `console.log`
* reorganized index for readability & code organization

### Added
* `app.registerProvider` method for providers
* `ejs` is now a dependency ([#234](https://github.com/koopjs/koop/pull/234))

@dmfenton @koopjs/dev

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -192,7 +88,73 @@ module.exports = function (config) {
...
       })
     }
+  }
+
+  /**
+   * express middleware setup
+   */
+
+  // save the koop log onto the app
+  app.log = koop.log
```
Comment (ngoldman) [inline index.js]: I don't think `app.log` is actually getting used anywhere. @dmfenton can I just take this out?

**Comment by chelm** [comment]
snazzzy :snowboarder:

File: index.js [hunk 2]
```diff
@@ -224,44 +186,103 @@ module.exports = function (config) {
...
+  /**
+   * registers koop providers
+   * exposes the provider's routes, controller, and model
+   * @param {object} provider - the provider to be registered
+   */
+  app.registerProvider = function (provider) {
+    app.services[provider.name] = provider
+
+    var model = new provider.model(koop) // eslint-disable-line
+    var controller = new provider.controller(model, koop.BaseController) // eslint-disable-line
```
Comment (ngoldman) [inline index.js]: @kneemer and I talked about this a bit... the model and controller in providers aren't actually constructors, they're just returning an object with a bunch of methods using the module pattern. Guessing @chelm read something like http://metaduck.com/08-module-pattern-inheritance.html and decided to go for module pattern over prototypal inheritance many moons ago.

What this means is we don't need the `new` keyword here as far as I know.

File: index.js [hunk 3]
```diff
@@ -224,44 +186,103 @@ module.exports = function (config) {
...
-  return app
+  /**
+   * registers a koop plugin
+   * Plugins can be any function that you want to have global access to
+   * within koop provider models
+   * @param {object} any koop plugin
+   */
+  app.registerPlugin = function (plugin) {
+    koop[plugin.name] = plugin
+  }
```
Comment (ngoldman) [inline index.js]: This method needs to be fleshed out if we want proper plugin support. I exposed `defaultRoutes` on the koop object above as a way of potentially letting plugins add routes to all providers for tiles, preview, etc.

**Comment by dmfenton** [comment]
I think the changes you made look good. But, I'm wondering, what's the test coverage on the areas you touched? In the past we haven't had good enough coverage and have ended up having to patch at least once after each refactor.

**Comment by ngoldman** [comment]
@dmfenton 90% of changes are just moving functions and var assignments around in `index.js` to make more sense. The only significant changes are

1. making `Files` accept options instead of koop, since the only options it needed are `log` and `config` (which is was grabbing off of `koop`).
1. removing `app.log = koop.log`. afaict it was never used anywhere.
1. using `koop.log` instead of `console.log` in the index.

The lib refactor was pretty big and pretty drastic (and I should have broken it up into smaller PRs). I'm pretty confident this one won't have any adverse effects. I can add some tests for the changes to the `register` process (it's acting the exact same for now, but making way in the future for any plugin or cache or provider to be registered via `app.register` and to self-identify via `name` and `type` properties).

**Comment by dmfenton** [comment]
@ngoldman That sounds good to me. It's just hard to tell how code has changed when a lot of things move around. If you've got this running well locally and we've got at least some tests in place, I'm good with this

**Comment by ngoldman** [comment]
I'm actually going to make one more change.

**Comment by ngoldman** [comment]
Okay I think this is good to go.

File: index.js [hunk 4]
```diff
@@ -1,122 +1,161 @@
...
+   * app methods
+   */
 
-  app.use(express.static(__dirname + '/public'))
+  /**
+   * registers koop providers, caches, and plugins
+   * @param {object} plugin - module to be registered
+   */
+  app.register = function (plugin) {
+    if (typeof plugin === 'undefined') return koop.log.error('Plugin undefined, skipping registration.')
```
Comment (dmfenton) [inline index.js]: why wouldn't we throw an error here?

File: index.js [hunk 5]
```diff
@@ -1,122 +1,161 @@
...
+    if (!plugin.name) return koop.log.error('Plugin missing name, skipping registration.')
```
Comment (dmfenton) [inline index.js]: same as above

File: index.js [hunk 6]
```diff
@@ -1,122 +1,161 @@
...
 
   // assigns a series of default routes; assumes
   app._bindDefaultRoutes = function (name, pattern, controller) {
-    var handler
-    for (handler in defaultRoutes) {
+    for (var handler in koop.defaultRoutes) {
       if (controller[handler]) {
-        defaultRoutes[handler].forEach(function (route) {
+        koop.defaultRoutes[handler].forEach(function (route) {
           app.get('/' + name + pattern + route, controller[handler])
```
Comment (dmfenton) [inline index.js]: aha, here's that pattern stuff

**Comment by dmfenton** [comment]
So just the quibble about throwing errors vs just logging them. Everything else is :ok_hand:

**Comment by ngoldman** [comment]
@dmfenton only reason I'm not throwing errors there is we weren't throwing before and I didn't want to break server deploys. I agree it should just throw if there are invalid plugins though. Shall we throw?

**Comment by dmfenton** [comment]
Definitely throw on invalid.

**Comment by ngoldman** [comment]
Ok this should be ready to go @dmfenton

**Comment by dmfenton** [comment]
:+1:

**Pull request merged**